crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs

```diff
@@ -431,3 +441,141 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
+                let segment_index = segment_header.segment_index();
+                let run_cache_sync_tx = run_cache_sync_tx.clone();
+
+                // Notify piece cache sync about new segment
+                if !run_cache_sync_tx.is_closed() {
+                    if let Err(err) = run_cache_sync_tx.send(segment_index).await {
+                        error!(
+                            "Failed to notify piece cache sync about new segment: {}",
+                            err
+                        );
+                    }
+                }
+
+                debug!(%segment_index, "Starting to process archived segment....");
+
+                for piece_index in segment_index.segment_piece_indexes() {
+                    let key = piece_index.hash().to_multihash().into();
+                    {
+                        if !piece_cache.lock().await.should_cache(&key) {
+                            trace!(%piece_index, ?key, "Piece key will not be included in the cache.");
+
+                            continue;
+                        }
+                    }
+
+                    trace!(%piece_index, ?key, "Piece key will be included in the cache.");
+
+                    // Segment notification will come earlier than node's local cache finishes its
+                    // initialization, so we need to wait for it.
+                    let mut retries_count = 0u16;
+                    'retry: loop {
+                        if retries_count >= GET_PIECE_MAX_RETRIES_COUNT {
+                            debug!(%piece_index, "Max retries number exceeded.");
+
+                            break 'retry;
+                        }
+
+                        retries_count += 1;
+
+                        let piece = node_client.piece(piece_index).await;
+
+                        match piece {
+                            Ok(Some(piece)) => {
+                                {
+                                    piece_cache.lock().await.add_piece(key, piece);
+                                }
+
+                                trace!(%piece_index, "Got piece for archived segment.");
+
+                                break 'retry;
+                            }
+                            Ok(None) => {
+                                debug!(%piece_index, "Can't get piece. Retrying...");
+
+                                sleep(Duration::from_secs(GET_PIECE_DELAY_IN_SECS)).await;
+                            }
+                            Err(err) => {
+                                warn!(
+                                    piece_index = ?piece_index,
+                                    err = ?err,
+                                    "Failed to get piece"
+                                );
+                            }
+                        }
+                    }
+                }
+
+                debug!(%segment_index, "Finished processing archived segment.");
+            }
```

Comment:
This is a little problematic IMHO. This assumes node doesn't produce segments faster than farmer processes them, which may or may not (during node sync) be true. When it is not true, notification stream above can exceed its size and drop RPC connection (IIRC), also pieces here will not be guaranteed to be available anymore.

I'm thinking we should probably re-introduce archiving acknowledgement here so we have a feedback loop from farmer back to the node.

Then we should only need one attempt and it should really be a hard error that piece is not there anymore right after it was archived.

Reply:
I restored the code we discussed previously from this commit https://github.com/subspace/subspace/commit/08b6ee400e10adbd7b1cdcf4df2b7eb0091fec65#diff-38362e1fdf3d3ede40d53ae7c45b7f7cb969a580b28c2adf80470b407cdcd8e3

I don't 100% follow the previous solution since it seems to have several drawbacks, the most important appears to be node hangs after farmers restart. It looks like we need to introduce some periodic cleanup for the obsolete acknowledgments. Also, the acknowledgment doesn't guarantee piece availability, we need to make a more complex workflow that seems overkill here to avoid retries. Did I understand you correctly? I welcome your comments here.

Reply:
Yes, it can hang on farmer disconnection if it happens in the middle of segment processing. This can be fixed though, you need to introduce a data structure that will be owned by the stream and will remove corresponding senders when dropped (it will be dropped when stream is dropped, which happens on disconnection). Not sure if jsonrpsee has a nicer API for this, didn't check.

Reply:
I found an easier way: remove the subscription state after the `pipe_from_stream` call.